rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+
+  The `callback_fn` will be invoked immediately after any of the three types
+  of events:
+    - The execution of an op under eager mode,
+    - The execution of a FuncGraph under eager mode,
+    - The creation of an op during graph construction (e.g., in
+      @tf.function-decorated Python functions.)
+
+  Args:
+    A callback_fn that has the following signature:
+      def callback_fn(op_type,
+                      inputs,
+                      attrs,
+                      outputs,
+                      op_name=None,
+                      graph=None):
+        # op_type: The type of the op. E.g., "MatMul".
+        #          For the special case of FuncGraph execution, op_type
+        #          takes the special value `tf.Graph`.
+        # inputs: (`tuple` of `Tensor`s) Input tensors to the op or the
+        #         FuncGraph.
+        #         In eager execution, these are `EagerTensor`s.
+        #         In graph construction, these are non-eager `Tensor`s
+        #         that form the inputs to the just-created op.
+        # attrs: The attributes of the op or FuncGraph of which the execution
+        #        or creation caused the current invocation of the callback.
+        # outputs: (`tuple of `Tensor`s) Output tensors from the op or
+        #          FuncGraph.
+        #          In eager execution, these are `EagerTensor`s.
+        #          In graph construction, these are non-eager `Tensor`s that
+        #          are the outputs of the just-created op.
+        # op_name: Name of the op or FuncGraph.
+        #          If the current invocation of the callback is due to the
+        #          eager execution of an op, this will be `None` as op names
+        #          are meaningless in eager execution.
+        #          If this invocation of the callback is due to the
+        #          eager execution of a FuncGraph, this will be the
+        #          internally-generated name of the FuncGraph.
+        #          In graph construction, this is the name of the op.
+        # graph: The graph that the op belongs to (if any).
+        #        In eager execution of an op, this is `None`.
+        #        In eager execution of a FuncGraph, the is the FuncGraph
+        #        object per se.
+        #        In graph construction, this is the op's containing graph.
+        #
+        # Return values:
+        #   This callback function is expected to return a `list` or `tuple`
+        #   of `Tensor`s, with its length matching `len(outputs)`.
+        #   In eager execution, these returned `Tensor`s should be
+        #   `EagerTensor`s. Their values will replace the original values of
+        #   `outputs` during downstream eager execution.
+        #   In graph construction, these returned `Tensor`s should be
+        #   non-eager `Tensor`s. Their values will replace the original
+        #   `outputs` during downstream graph construction.
```

Comment:
Can the output dtypes be different?

Reply:
I don't see why this API should prohibit changing the dtype. But beware that if you override the dtype (e.g., float32 to int64), downstream graph construction or eager execution may likely error out with a dtype-not-supported error.